README.md
# Gyazo for Linux

[![Circle CI](https://circleci.com/gh/gyazo/Gyazo-for-Linux.svg?style=svg)](https://circleci.com/gh/gyazo/Gyazo-for-Linux)

https://gyazo.com/


### Install

#### Debian/Ubuntu package

[Download a deb file](https://github.com/DarkminecrafterHD/Gyazo-for-Linux/releases).

Install Ruby, ImageMagick and Xdotool before installing Gyazo.

    $ sudo apt-get install ruby imagemagick xdotool
    $ sudo dpkg -i gyazo_XXX_all.deb


#### apt-get (Oficial version) [Git here](https://github.com/gyazo/Gyazo-for-Linux)

    $ curl -s https://packagecloud.io/install/repositories/gyazo/gyazo-for-linux/script.deb.sh | sudo bash
    $ sudo apt-get install gyazo


#### How to add a Gyazo icon to Ubuntu Unity Launcher

1. Open Unity Dash
2. Search "Gyazo"
3. Drag the Gyazo icon and drop into the launcher

#### :warning: Trouble Shooting: Failed to install

- Please refer to [this issue](https://github.com/gyazo/Gyazo-for-Linux/issues/35).
  - Install `deb` package, write `os` to `ubuntu` and `version` to `trusty`
  - Install `rpm` package, write `os` to `el` and `version` to `6`

#### yum (Oficial version) [Git here](https://github.com/gyazo/Gyazo-for-Linux)

    $ curl -s https://packagecloud.io/install/repositories/gyazo/gyazo-for-linux/script.rpm.sh | sudo bash
    $ sudo yum install gyazo

### How to change screenshot-tool

Gyazo use `import`(imagemagick) command by default.
If you have some trouble on screenshot such as cannot take correct area, take broken image...,try to change screenshot-tool by this way.

Revise `/src/gyazo.rb`[here](https://github.com/gyazo/Gyazo-for-Linux/blob/3451db33631a0732097ed1cfaa87326672695a27/src/gyazo.rb#L24
) like examples.

- scrot

```diff
if imagefile && File.exist?(imagefile) then
  system "convert '#{imagefile}' '#{tmpfile}'"
else
-  system "import '#{tmpfile}'"
+  system "scrot -s '#{tmpfile}'"
end

```

- gnome-screenshot

```diff
if imagefile && File.exist?(imagefile) then
  system "convert '#{imagefile}' '#{tmpfile}'"
else
-  system "import '#{tmpfile}'"
+  system "gnome-screenshot -a -f '#{tmpfile}'"
end
```

- other screenshot tools

https://wiki.archlinux.org/index.php/Taking_a_screenshot

#### Contributions
Pull requests are welcome.

Gyazo for Linux is maintained by Nota Inc. But the development is not as active as that of Windows / Mac versions. We can't fix a reported problem soon.

If you want a problem fixed soon, we recommend you fix it by yourself.

We would be glad if you send a PR to this repository.

#### How to release

- Create Pull Request from `master` to `release`
    - Please mark version with `git tag`
- Build package and push to packagecloud by CircleCI when it merged

### License

Copyright (c) 2015 Nota Inc.

This software is licensed under the GPL



debian/changelog
gyazo (1.3.1-darkangeel-hd) unstable; urgency=low
	
  * Fixed Copy link to clipboard.

 -- DarkminecraferHD <DarkminecrafterHD@users.noreply.github.com>  Wed, 10 Feb 2016 16:17:01 +0900

gyazo (1.3-darkangeel-hd) unstable; urgency=low
	
  * Add gnome integration in another executable.
  * Fixed Copy link to clipboard.

 -- DarkminecraferHD <DarkminecrafterHD@users.noreply.github.com>  Wed, 10 Feb 2016 00:28:01 +0900

gyazo (1.2-darkangeel-hd-patch-1) unstable; urgency=low
	
  * Fixed Copy link to clipboard.

 -- NoNameA 774 <DarkminecrafterHD@users.noreply.github.com>  Wed, 10 Feb 2016 00:20:01 +0900

gyazo (1.1) unstable; urgency=low
	
  * Use https.

 -- NoNameA 774 <nonamea774@gmail.com>  Thu, 2 Jul 2015 13:53:01 +0900

gyazo (1.0-1) unstable; urgency=low

  * Initial release.

 -- kambara <kambara@sappari.org>  Wed, 19 Oct 2011 21:16:29 +0900


src/gyazo.rb
#!/usr/bin/env ruby

# setting
browser_cmd = 'xdg-open'
clipboard_cmd = 'xclip -selection c -i'

require 'net/http'
require 'open3'
require 'openssl'
require 'json'

# get id
idfile = ENV['HOME'] + "/.gyazo.id"

id = ''
if File.exist?(idfile) then
  id = File.read(idfile).chomp
end

# get active window name

active_window_id = `xprop -root | grep "_NET_ACTIVE_WINDOW(WINDOW)" | cut -d ' ' -f 5`.chomp
out, err, status = Open3.capture3 "xwininfo -id #{active_window_id} | grep \"xwininfo: Window id: \"|sed \"s/xwininfo: Window id: #{active_window_id}//\""
active_window_name = out.chomp
out, err, status = Open3.capture3 "xprop -id #{active_window_id} | grep \"_NET_WM_PID(CARDINAL)\" | sed s/_NET_WM_PID\\(CARDINAL\\)\\ =\\ //"

pid = out.chomp

application_name = `ps -p #{pid} -o comm=`.chomp
# capture png file
tmpfile = "/tmp/image_upload#{$$}.png"
imagefile = ARGV[0]

if imagefile && File.exist?(imagefile) then
  system "convert '#{imagefile}' '#{tmpfile}'"
else
  system "import '#{tmpfile}'"
end

if !File.exist?(tmpfile) then
  exit
end

imagedata = File.read(tmpfile)
File.delete(tmpfile)

xuri = ""
if application_name =~ /(chrom(ium|e)|firefox|iceweasel)/
  xuri = `xdotool windowfocus #{active_window_id}; xdotool key "ctrl+l"; xdotool key "ctrl+c"; xclip -o`
end


# upload
boundary = '----BOUNDARYBOUNDARY----'

# endpoint https://gyazo.com/api/docs
HOST = 'upload.gyazo.com'
CGI = '/upload.cgi'
UA   = 'Gyazo/1.2'

metadata = JSON.generate({
  app: active_window_name,
  title: active_window_name,
  url: xuri,
  note: "#{active_window_name}\n#{xuri}"
})

data = <<EOF
--#{boundary}\r
content-disposition: form-data; name="metadata"\r
\r
#{metadata}\r
--#{boundary}\r
content-disposition: form-data; name="id"\r
\r
#{id}\r
--#{boundary}\r
content-disposition: form-data; name="imagedata"; filename="gyazo.com"\r
\r
#{imagedata}\r
--#{boundary}--\r
EOF

header ={
  'Content-Length' => data.length.to_s,
  'Content-type' => "multipart/form-data; boundary=#{boundary}",
  'User-Agent' => UA
}

env = ENV['http_proxy']
if env then
  uri = URI(env)
  proxy_host, proxy_port = uri.host, uri.port
else
  proxy_host, proxy_port = nil, nil
end
https = Net::HTTP::Proxy(proxy_host, proxy_port).new(HOST,443)
https.use_ssl = true
https.verify_mode = OpenSSL::SSL::VERIFY_PEER
https.verify_depth = 5
https.start{
  res = https.post(CGI,data,header)
  url = res.response.body
  puts url
  system "#{browser_cmd} '#{url}'"
  system "echo -n '#{url}' | #{clipboard_cmd}"

  # save id
  newid = res.response['X-Gyazo-Id']
  if newid and newid != "" then
    if !File.exist?(File.dirname(idfile)) then
      Dir.mkdir(File.dirname(idfile))
    end
    if File.exist?(idfile) then
      File.rename(idfile, idfile+Time.new.strftime("_%Y%m%d%H%M%S.bak"))
    end
    File.open(idfile,"w").print(newid)
  end
}


src/gyazo-gnome.rb
#!/usr/bin/env ruby

# setting
browser_cmd = 'xdg-open'
clipboard_cmd = 'xclip -selection c -i'

require 'net/http'
require 'open3'
require 'openssl'
require 'json'

# get id
idfile = ENV['HOME'] + "/.gyazo.id"

id = ''
if File.exist?(idfile) then
  id = File.read(idfile).chomp
end

# get active window name

active_window_id = `xprop -root | grep "_NET_ACTIVE_WINDOW(WINDOW)" | cut -d ' ' -f 5`.chomp
out, err, status = Open3.capture3 "xwininfo -id #{active_window_id} | grep \"xwininfo: Window id: \"|sed \"s/xwininfo: Window id: #{active_window_id}//\""
active_window_name = out.chomp
out, err, status = Open3.capture3 "xprop -id #{active_window_id} | grep \"_NET_WM_PID(CARDINAL)\" | sed s/_NET_WM_PID\\(CARDINAL\\)\\ =\\ //"

pid = out.chomp

application_name = `ps -p #{pid} -o comm=`.chomp
# capture png file
tmpfile = "/tmp/image_upload#{$$}.png"
imagefile = ARGV[0]

if imagefile && File.exist?(imagefile) then
  system "convert '#{imagefile}' '#{tmpfile}'"
else
  system "gnome-screenshot -a -f '#{tmpfile}'"
end

if !File.exist?(tmpfile) then
  exit
end

imagedata = File.read(tmpfile)
File.delete(tmpfile)

xuri = ""
if application_name =~ /(chrom(ium|e)|firefox|iceweasel)/
  xuri = `xdotool windowfocus #{active_window_id}; xdotool key "ctrl+l"; xdotool key "ctrl+c"; xclip -o`
end


# upload
boundary = '----BOUNDARYBOUNDARY----'

# endpoint https://gyazo.com/api/docs
HOST = 'upload.gyazo.com'
CGI = '/upload.cgi'
UA   = 'Gyazo/1.2'

metadata = JSON.generate({
  app: active_window_name,
  title: active_window_name,
  url: xuri,
  note: "#{active_window_name}\n#{xuri}"
})

data = <<EOF
--#{boundary}\r
content-disposition: form-data; name="metadata"\r
\r
#{metadata}\r
--#{boundary}\r
content-disposition: form-data; name="id"\r
\r
#{id}\r
--#{boundary}\r
content-disposition: form-data; name="imagedata"; filename="gyazo.com"\r
\r
#{imagedata}\r
--#{boundary}--\r
EOF

header ={
  'Content-Length' => data.length.to_s,
  'Content-type' => "multipart/form-data; boundary=#{boundary}",
  'User-Agent' => UA
}

env = ENV['http_proxy']
if env then
  uri = URI(env)
  proxy_host, proxy_port = uri.host, uri.port
else
  proxy_host, proxy_port = nil, nil
end
https = Net::HTTP::Proxy(proxy_host, proxy_port).new(HOST,443)
https.use_ssl = true
https.verify_mode = OpenSSL::SSL::VERIFY_PEER
https.verify_depth = 5
https.start{
  res = https.post(CGI,data,header)
  url = res.response.body
  puts url
  system "#{browser_cmd} '#{url}'"
  system "echo -n '#{url}' | #{clipboard_cmd}"

  # save id
  newid = res.response['X-Gyazo-Id']
  if newid and newid != "" then
    if !File.exist?(File.dirname(idfile)) then
      Dir.mkdir(File.dirname(idfile))
    end
    if File.exist?(idfile) then
      File.rename(idfile, idfile+Time.new.strftime("_%Y%m%d%H%M%S.bak"))
    end
    File.open(idfile,"w").print(newid)
  end
}
